tokio/src/task/coop/mod.rs

```diff
@@ -275,27 +288,100 @@ cfg_coop! {
+    /// Decrements the task budget and returns [`Poll::Pending`] if the budget is depleted.
+    /// This indicates that the task should yield to the scheduler. Otherwise, returns
+    /// [`RestoreOnPending`] which can be used to commit the budget consumption.
     ///
-    /// When you call this method, the current budget is decremented. However, to ensure that
-    /// progress is made every time a task is polled, the budget is automatically restored to its
-    /// former value if the returned `RestoreOnPending` is dropped. It is the caller's
-    /// responsibility to call `RestoreOnPending::made_progress` if it made progress, to ensure
-    /// that the budget empties appropriately.
+    /// The returned [`RestoreOnPending`] will revert the budget to its former
+    /// value when dropped unless [`RestoreOnPending::made_progress`]
+    /// is called. It is the caller's responsibility to do so when it _was_ able to
+    /// make progress after the call to [`poll_proceed`].
+    /// Restoring the budget automatically ensures the task can try to make progress in some other
+    /// way.
     ///
-    /// Note that `RestoreOnPending` restores the budget **as it was before `poll_proceed`**.
-    /// Therefore, if the budget is _further_ adjusted between when `poll_proceed` returns and
-    /// `RestRestoreOnPending` is dropped, those adjustments are erased unless the caller indicates
+    /// Note that [`RestoreOnPending`] restores the budget **as it was before [`poll_proceed`]**.
+    /// Therefore, if the budget is _further_ adjusted between when [`poll_proceed`] returns and
+    /// [`RestoreOnPending`] is dropped, those adjustments are erased unless the caller indicates
     /// that progress was made.
+    ///
+    /// # Examples
+    ///
+    /// This example shows a simple countdown latch that uses [`poll_proceed`] to participate in
+    /// cooperative scheduling.
+    ///
+    /// ```
+    /// use std::future::{Future};
+    /// use std::pin::Pin;
+    /// use std::task::{ready, Context, Poll, Waker};
+    /// use tokio::task::coop;
+    ///
+    /// struct CountdownLatch<T> {
+    ///     counter: usize,
+    ///     value: Option<T>,
+    ///     waker: Option<Waker>
+    /// }
+    ///
+    /// impl<T> CountdownLatch<T> {
+    ///     fn new(value: T, count: usize) -> Self {
+    ///         CountdownLatch {
+    ///             counter: count,
+    ///             value: Some(value),
+    ///             waker: None
+    ///         }
+    ///     }
+    ///     fn count_down(&mut self) {
+    ///         if self.counter <= 0 {
+    ///             return;
+    ///         }
+    ///
+    ///         self.counter -= 1;
+    ///         if self.counter == 0 {
+    ///             if let Some(w) = self.waker.take() {
+    ///                 w.wake();
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    ///
+    /// impl<T> Future for CountdownLatch<T> where T: Unpin {
```

Comment:
Unpin is also not needed here.

Reply:
You can add
```
impl<T> Unpin for CountdownLatch<T> {}
```

Reply:
When you write `impl<T> Unpin for CountdownLatch<T> {}`, that's a declaration that `CountdownLatch<T>` is never self-referential. It makes pinning it a no-op.

Reply:
> Just because you pin a struct does not mean you pin the fields.

After 20+ years of Java, it takes some time to become fluent in this stuff. I'll get there eventually 😄.

My rationale was that you would want to prevent implicitly moving some !Unpin type by moving the CountdownLatch itself. If a user of this example CountdownLatch wants to put something unmovable in it, they should pin the value themselves and use something like a `Pin<Box<T>>` as the value in the latch then, right?